### PR TITLE
F-498: add support for `passwd` and `random_passwd` arguments

### LINKED ISSUE #498: Add `passwd` and `random_passwd` arguments for opennebula_virtual_machine and opennebula_template `graphics` sections
### Description

These resources currently doesn't support `passwd` or `random_passwd` arguments inside `graphics` section:
```
│ Error: Unsupported argument
│ 
│   on ../../_modules/instance/instance.tf line 33, in resource "opennebula_virtual_machine" "instance":
│   33:     passwd = substr(base64encode(random_string.password.result), 0, 8)
│ 
│ An argument named "passwd" is not expected here.
```

### New or affected resources and data sources

opennebula_virtual_machine
opennebula_template

### Potential terraform configuration

```HCL
resource "opennebula_virtual_machine" "example" {
  count = 2

  name        = "virtual-machine-${count.index}"
  description = "VM"
  cpu         = 1
  vcpu        = 1
  memory      = 1024
  group       = "terraform"
  permissions = "660"

  context = {
    NETWORK      = "YES"
    HOSTNAME     = "$NAME"
    START_SCRIPT = "yum upgrade"
  }

  graphics {
    type   = "VNC"
    listen = "0.0.0.0"
    keymap = "fr"
    passwd = substr(base64encode(random_string.password.result), 0, 8)
  }

  os {
    arch = "x86_64"
    boot = "disk0"
  }

  disk {
    image_id = opennebula_image.example.id
    size     = 10000
    target   = "vda"
    driver   = "qcow2"
  }

  on_disk_change = "RECREATE"

  nic {
    model           = "virtio"
    network_id      = var.vnetid
    security_groups = [opennebula_security_group.example.id]
  }

  vmgroup {
    vmgroup_id = 42
    role       = "vmgroup-role"
  }

  sched_requirements = "FREE_CPU > 60"

  tags = {
    environment = "example"
  }

  template_section {
   name = "example"
   elements = {
      key1 = "value1"
   }
  }
}
```

### References

_No response_

### PR DESCRIPTION
**Pull request by grossibr**
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for PR followers and do not help prioritize the request

### Description

Adds support for `passwd` and `random_passwd` arguments for `graphics` section for VNC so we avoid the following:
```
│ Error: Unsupported argument
│
│ on ../../_modules/instance/instance.tf line 33, in resource "opennebula_virtual_machine" "instance":
│ 33: passwd = substr(base64encode(random_string.password.result), 0, 8)
│
│ An argument named "passwd" is not expected here.
<!--- Please leave a helpful description of the PR here. --->
```
### References

#498: Add `passwd` and `random_passwd` arguments for opennebula_virtual_machine and opennebula_template `graphics` sections

### New or Affected Resource(s)

- `opennebula_virtual_machine`
- `opennebula_template`

### Checklist

- [x] I have created an issue and I have mentioned it in `References`
- [x] My code follows the style guidelines of this project (use `go fmt`)
- [x] My changes generate no new warnings or errors
- [ ] I have updated the unit tests and they pass succesfuly - NA
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated the documentation (if needed)
- [x] I have updated the changelog file

### CODE REVIEW

File: opennebula/shared_schemas.go [hunk 1]
```diff
@@ -594,6 +602,10 @@ func addGraphic(tpl *vm.Template, graphics []interface{}) {
 				tpl.AddIOGraphic(vmk.Port, v.(string))
 			case "keymap":
 				tpl.AddIOGraphic(vmk.Keymap, v.(string))
+			case "passwd":
+				tpl.AddIOGraphic(vmk.Passwd, v.(string))
+			case "random_passwd":
+				tpl.AddIOGraphic(vmk.Passwd, v.(bool))
```
Comment (sefimov-axc) [inline opennebula/shared_schemas.go]: vmk.RandomPassword

File: opennebula/shared_schemas.go [hunk 2]
```diff
@@ -869,6 +881,8 @@ func flattenTemplate(d *schema.ResourceData, inheritedVectors map[string]interfa
 	port, _ := vmTemplate.GetIOGraphic(vmk.Port)
 	t, _ := vmTemplate.GetIOGraphic(vmk.GraphicType)
 	keymap, _ := vmTemplate.GetIOGraphic(vmk.Keymap)
+	passwd, _ := vmTemplate.GetIOGraphic(vmk.Passwd)
+	random_passwd, _ := vmTemplate.GetIOGraphic(vmk.RandomPasswd)
```
Comment (sefimov-axc) [inline opennebula/shared_schemas.go]: vmk.RandomPassword

Comment (grossibr) [reply opennebula/shared_schemas.go]: thanks

File: opennebula/shared_schemas.go (hunk 1 above)
Comment (grossibr) [reply opennebula/shared_schemas.go]: thanks

**Comment by grossibr** [comment]
@frousselet may I ask you to approve the flows once more?

**Comment by frousselet** [comment]
Done, @axbgrossi :)

**Comment by grossibr** [comment]
@frousselet once more if you're available, please

**Comment by grossibr** [comment]
`CI / Test OpenNebula 5.12` error seems unrelated to the changes in this PR, please resume with review

**Review by treywelsh — APPROVED** [review]

**Comment by grossibr** [comment]
@treywelsh @frousselet will this be merged by you maintainers when the time comes?

**Comment by grossibr** [comment]
@frousselet please consider reviewing this PR so we can finally close that issue

**Review by frousselet — CHANGES_REQUESTED** [review]
Hello @axbgrossi,

Thanks for your work.

If I understand the documentation, random password should be a boolean (then converted in a `'YES'` or a `'NO'` string by the provider).
The way you did is a good start but it might be too permissive.

Also, both `passwd` and `random_passwd` should conflict with each other.

**Comment by grossibr** [comment]
@frousselet could you please approve the workflows?

**Comment by grossibr** [comment]
@frousselet would you be kind to approve the workflows once more?

File: opennebula/shared_schemas.go [hunk 3]
```diff
@@ -929,11 +944,17 @@ func flattenTemplate(d *schema.ResourceData, inheritedVectors map[string]interfa
 
 	// Set graphics to resource
 	if port != "" {
+		// passwd and random_passwd can't coexist
+		if passwd != "" && random_passwd != "" {
+			return fmt.Errorf("The arguments 'passwd' and 'random_passwd' can't coexist")
```
Comment (frousselet) [inline opennebula/shared_schemas.go]: You probably would prefer to use the native SDK option `ConflictsWith`.

[Example](https://github.com/OpenNebula/terraform-provider-opennebula/blob/6c416689b850d4738246291902cc0e164defad23/opennebula/resource_opennebula_image.go#L105)

File: website/docs/r/template.html.markdown [hunk 4]
```diff
@@ -114,6 +114,8 @@ The following arguments are supported:
 * `listen` - (Required) Binding address.
 * `port` - (Optional) Binding Port.
 * `keymap` - (Optional) Keyboard mapping.
+* `passwd` - (Optional) VNC's password.
+* `random_passwd` - (Optional) Randomized VNC's password.
```
Comment (frousselet) [inline website/docs/r/template.html.markdown]: Conflicts should be mentioned on the documentation

File: website/docs/r/virtual_machine.html.markdown [hunk 5]
```diff
@@ -117,6 +117,8 @@ The following arguments are supported:
 * `listen` - (Required) Binding address.
 * `port` - (Optional) Binding Port.
 * `keymap` - (Optional) Keyboard mapping.
+* `passwd` - (Optional) VNC's password.
+* `random_passwd` - (Optional) Randomized VNC's password.
```
Comment (frousselet) [inline website/docs/r/virtual_machine.html.markdown]: Conflicts should be mentioned on the documentation

**Review by frousselet — CHANGES_REQUESTED** [review]
Hi @axbgrossi,

Thanks for your work!

There is a SDK native `ConflictsWith` argument you can use to natively deal with conflicting attributes.

**Comment by grossibr** [comment]
@frousselet would you please approve the flows once more?

File: website/docs/r/virtual_machine.html.markdown (hunk 5 above)
Comment (grossibr) [reply website/docs/r/virtual_machine.html.markdown]: done

File: website/docs/r/template.html.markdown (hunk 4 above)
Comment (grossibr) [reply website/docs/r/template.html.markdown]: done

File: opennebula/shared_schemas.go (hunk 3 above)
Comment (grossibr) [reply opennebula/shared_schemas.go]: had issues before so I switched the approach, trying this way one more time

Comment (frousselet) [reply opennebula/shared_schemas.go]: Many thanks. Do not hesitate to ask for help if you need some.

**Comment by grossibr** [comment]
@frousselet once more please

**Comment by treywelsh** [comment]
Seems due to the ConflictsWith: their values targets nested Graphics attributes, not attributes at the root of the schema.
So try something "graphics.0.passwd" and "graphics.0.random_passwd". It's ok because Graphics is a typelist with a maximum of 1 element

**Comment by grossibr** [comment]
> Seems due to the ConflictsWith: their values targets nested Graphics attributes, not attributes at the root of the schema. So try something "graphics.0.passwd" and "graphics.0.random_passwd". It's ok because Graphics is a typelist with a maximum of 1 element

done, @treywelsh or @frousselet could you please approve the flows?

**Review by frousselet — APPROVED** [review]

**Review by frousselet — APPROVED** [review]

**Comment by frousselet** [comment]
Hello @axbgrossi,

Thanks a lot for your work.
Could you rebase with master and solve merge conflicts please?

Then I will be able to merge and include it in the `1.4.0` release.

**Pull request merged**